### DISCUSSION TITLE
Passing large amounts of data to pages via context not recommended?

### DISCUSSION THREAD

**Discussion by samajammin** [gatsbyjs/gatsby#30765] [Help]
Hey folks, I came across this recommendation in the Scaling docs:
https://www.gatsbyjs.com/docs/scaling-issues/#switch-off-type-inference-for-sitepagecontext

> When using the createPages API to pass large amounts of data to pages via context (which is generally not recommended), Gatsby’s type inference can become slow.

I'm curious why this approach is not recommended. We use a community plugin for internationalization:
https://github.com/wiziple/gatsby-plugin-intl/

This plugin passes message strings for all copy on our site (1K > message strings) for every page (for >1K pages) into context. Is this something we should avoid?

I found a separate discussion on context size (https://github.com/gatsbyjs/gatsby/discussions/30675) but we have never seen that warning before:
```
warning The size of at least one page context chunk exceeded 500kb, which could lead to degraded performance. Consider putting less data in the page context.
```
So I'm hoping we haven't hit that point of concern.

Appreciate any advice on this. Would be great if we could add more explanation to these docs, which I'd be happy to do once I understand the "why". Thanks in advance!

**Comment by LekoArts** [comment]
Here's the page that it should link to: https://www.gatsbyjs.com/docs/creating-and-modifying-pages/#performance-implications
I've opened a PR: https://github.com/gatsbyjs/gatsby/pull/30778

> This plugin passes message strings for all copy on our site (1K > message strings) for every page (for >1K pages) into context. Is this something we should avoid?

Yes! It means that any page gets all the data.

I guess the most relevant discussion on this is here: https://github.com/gatsbyjs/gatsby/discussions/13806

What I did in the past was loading the json files and defining the path to it via webpack define plugin:
https://github.com/gatsbyjs/themes/blob/master/packages/gatsby-theme-i18n-react-intl/gatsby-node.js
https://github.com/gatsbyjs/themes/blob/master/packages/gatsby-theme-i18n-react-intl/src/wrap-page-element.js
